# Can you solve the human cannonball riddle? - Alex Rosenthal
## [Source](https://www.youtube.com/watch?v=o4MpBV4F3qs&list=WL)

### [Glitched Failure Video](https://youtu.be/In7NyqE81C8)

## Approach

1. Generate all possible Cannon states
1. Create Cannon ruleset
1. Iterate over all possible Cannon states
1. Analyze all states that pass ruleset

## Code I wish I had

```python
ruleset = [rule_1, rule_2, rule_3]
cannon_gen = create_cannon_state_generator()

solutions = []
for cannon_state in cannon_gen:
    passes = True
    for rule in ruleset:
        passes &= rule(cannon_state)
        if not passes:
            break

    if passes:
        solutions.append(cannon_state)
```

### `create_cannon_state_generator`

Generators can be used to make efficient use of space

In [1]:
# example of generator
def fib_gen():
    yield 1
    yield 1
    
    prev_vals = [1,1]
    while True:
        next_val = sum(prev_vals)
        prev_vals.pop(0)
        prev_vals.append(next_val)

        yield next_val

In [19]:
fb = fib_gen()

In [20]:
for i in fb:
    print(i)

1
1
2
3
5
8
13
21
34
55
89
144
233
377
610
987
1597
2584
4181
6765
10946
17711
28657
46368
75025
121393
196418
317811
514229
832040
1346269
2178309
3524578
5702887
9227465
14930352
24157817
39088169
63245986
102334155
165580141
267914296
433494437
701408733
1134903170
1836311903
2971215073
4807526976
7778742049
12586269025
20365011074
32951280099
53316291173
86267571272
139583862445
225851433717
365435296162
591286729879
956722026041
1548008755920
2504730781961
4052739537881
6557470319842
10610209857723
17167680177565
27777890035288
44945570212853
72723460248141
117669030460994
190392490709135
308061521170129
498454011879264
806515533049393
1304969544928657
2111485077978050
3416454622906707
5527939700884757
8944394323791464
14472334024676221
23416728348467685
37889062373143906
61305790721611591
99194853094755497
160500643816367088
259695496911122585
420196140727489673
679891637638612258
1100087778366101931
1779979416004714189
2880067194370816120
4660046610375530309
7540113804746346429


IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)




409079342168043908827205263103673495288925050163227508148465179505376738947630245434704358741879551893587981873537323724200566187478233426808109788866205835385082415438775973819467730078853211413259691666328215453273152704781822095617379223866873820618186905922125996236786182752629012607129857186122697922302293689660488384100932745999592904589171966199644174701151983794099575974801658563064124865281865252536233130764522878175997941099779766585474075277138800126285614456329819194452799124017317568308127513525192216268510422215988286910374664753583019879454649305627336963454626740039360193478895068484576406096758438628248496363403337931199206670919084663000936019981505985699221956088123083338918417402635515093782324450000619200267726432824226705656786943272154759098688195913091846620666882481442199091898804344500457084149980971871897722447269542983105740793727124842525712556822730711324736933410603751755271451442234612329933116905596053579695493430672177486367673562385909190949635004848

KeyboardInterrupt: 

In [25]:
from itertools import product


In [30]:
list(product([1,2,3], repeat=8))

[(1, 1, 1, 1, 1, 1, 1, 1),
 (1, 1, 1, 1, 1, 1, 1, 2),
 (1, 1, 1, 1, 1, 1, 1, 3),
 (1, 1, 1, 1, 1, 1, 2, 1),
 (1, 1, 1, 1, 1, 1, 2, 2),
 (1, 1, 1, 1, 1, 1, 2, 3),
 (1, 1, 1, 1, 1, 1, 3, 1),
 (1, 1, 1, 1, 1, 1, 3, 2),
 (1, 1, 1, 1, 1, 1, 3, 3),
 (1, 1, 1, 1, 1, 2, 1, 1),
 (1, 1, 1, 1, 1, 2, 1, 2),
 (1, 1, 1, 1, 1, 2, 1, 3),
 (1, 1, 1, 1, 1, 2, 2, 1),
 (1, 1, 1, 1, 1, 2, 2, 2),
 (1, 1, 1, 1, 1, 2, 2, 3),
 (1, 1, 1, 1, 1, 2, 3, 1),
 (1, 1, 1, 1, 1, 2, 3, 2),
 (1, 1, 1, 1, 1, 2, 3, 3),
 (1, 1, 1, 1, 1, 3, 1, 1),
 (1, 1, 1, 1, 1, 3, 1, 2),
 (1, 1, 1, 1, 1, 3, 1, 3),
 (1, 1, 1, 1, 1, 3, 2, 1),
 (1, 1, 1, 1, 1, 3, 2, 2),
 (1, 1, 1, 1, 1, 3, 2, 3),
 (1, 1, 1, 1, 1, 3, 3, 1),
 (1, 1, 1, 1, 1, 3, 3, 2),
 (1, 1, 1, 1, 1, 3, 3, 3),
 (1, 1, 1, 1, 2, 1, 1, 1),
 (1, 1, 1, 1, 2, 1, 1, 2),
 (1, 1, 1, 1, 2, 1, 1, 3),
 (1, 1, 1, 1, 2, 1, 2, 1),
 (1, 1, 1, 1, 2, 1, 2, 2),
 (1, 1, 1, 1, 2, 1, 2, 3),
 (1, 1, 1, 1, 2, 1, 3, 1),
 (1, 1, 1, 1, 2, 1, 3, 2),
 (1, 1, 1, 1, 2, 1, 3, 3),
 (1, 1, 1, 1, 2, 2, 1, 1),
 

In [31]:
from itertools import product

def create_cannon_state_generator(layers, min_max_range):
    min_val, max_val = min_max_range
    for cannon_state in product(product(range(min_val, max_val+1), repeat=8), repeat=layers):
        yield cannon_state
    return

In [32]:
csg = create_cannon_state_generator(layers = 2, min_max_range = (1,3))

for _ in range(10):
    print(next(csg))

((1, 1, 1, 1, 1, 1, 1, 1), (1, 1, 1, 1, 1, 1, 1, 1))
((1, 1, 1, 1, 1, 1, 1, 1), (1, 1, 1, 1, 1, 1, 1, 2))
((1, 1, 1, 1, 1, 1, 1, 1), (1, 1, 1, 1, 1, 1, 1, 3))
((1, 1, 1, 1, 1, 1, 1, 1), (1, 1, 1, 1, 1, 1, 2, 1))
((1, 1, 1, 1, 1, 1, 1, 1), (1, 1, 1, 1, 1, 1, 2, 2))
((1, 1, 1, 1, 1, 1, 1, 1), (1, 1, 1, 1, 1, 1, 2, 3))
((1, 1, 1, 1, 1, 1, 1, 1), (1, 1, 1, 1, 1, 1, 3, 1))
((1, 1, 1, 1, 1, 1, 1, 1), (1, 1, 1, 1, 1, 1, 3, 2))
((1, 1, 1, 1, 1, 1, 1, 1), (1, 1, 1, 1, 1, 1, 3, 3))
((1, 1, 1, 1, 1, 1, 1, 1), (1, 1, 1, 1, 1, 2, 1, 1))


43,040,160 possible cannon states!

### Convert flat layer list into 3D orientation

Indices for each side


```python
layer = [0, 1, 2, 3, 4, 5, 6, 7]

         north
     | 0 | 1 | 2 |
west | 3 |   | 4 | east
     | 5 | 6 | 7 |
         south

```

In [144]:
# Create mapping of flat cannon state layers
# contains the indices for each side
layer_map = {
    'north': [2,1,0],
    'south': [5,6,7],
    'west' : [0,3,5],
    'east' : [7,4,2],
}

In [238]:
from colorama import Fore, Style

def vis_cannon(cannon_state):
    layer_1 = [f'{Fore.GREEN}{val}{Style.RESET_ALL}'.strip() for val in cannon_state[0]] # lower
    layer_2 = [f'{Fore.RED}{val}{Style.RESET_ALL}'.strip() for val in cannon_state[1]] # upper

    layer_map = {
        'north': [2,1,0],
        'south': [5,6,7],
        'west' : [0,3,5],
        'east' : [7,4,2],
    }
    
    lower = \
    """
          LOWER
            n
      | {} | {} | {} |
    w | {} |   | {} | e
      | {} | {} | {} |
            s
    """.format(*layer_1)
    
    upper = \
    """
          UPPER
            n
        | {} | {} | {} |
    w | {} |   | {} | e
        | {} | {} | {} |
            s
    """.format(*layer_2)
        
    north = \
    """
        NORTH
        upper
      | {} | {} | {} |
    | {} | {} | {} |
          lower
    """.format(*[layer_2[i] for i in layer_map['north']], *[layer_1[i] for i in layer_map['north']])
       
    south = \
    """
    SOUTH
    upper
    | {} | {} | {} |
    | {} | {} | {} |
      lower
    """.format(*[layer_2[i] for i in layer_map['south']], *[layer_1[i] for i in layer_map['south']])
       
    east = \
    """
  EAST
  upper
      | {} | {} | {} |
      | {} | {} | {} |
    lower
    """.format(*[layer_2[i] for i in layer_map['east']], *[layer_1[i] for i in layer_map['east']])
       
    west = \
    """
WEST
upper
      | {} | {} | {} |
      | {} | {} | {} |
  lower
    """.format(*[layer_2[i] for i in layer_map['west']], *[layer_1[i] for i in layer_map['west']])
       
    # PRINTING #
    fixed_width = 21
    def set_width(string):
        return f"{string:<{fixed_width}}"

    all_sides = [lower, upper, north, south, east, west]
    for combined_str in zip(*[side.split('\n') for side in all_sides]):
        print(''.join([set_width(string) for string in combined_str]))

In [239]:
# Showing lower and upper indices
vis_cannon([range(8), range(8)])

                                                                                                                              
          LOWER                UPPER              NORTH            SOUTH              EAST               WEST                 
            n                    n                upper            upper              upper              upper                
      | 0 | 1 | 2 |        | 0 | 1 | 2 |      | 2 | 1 | 0 |    | 5 | 6 | 7 |      | 7 | 4 | 2 |      | 0 | 3 | 5 |
    w | 3 |   | 4 | e    w | 3 |   | 4 | e    | 2 | 1 | 0 |    | 5 | 6 | 7 |      | 7 | 4 | 2 |      | 0 | 3 | 5 |
      | 5 | 6 | 7 |        | 5 | 6 | 7 |          lower            lower              lower              lower              
            s                    s                                                                                            


### What should a "rule" look like?

- Will be a function corresponding to each given rule
- Will take all Cannon layers as nested "flat" list representation
- Returns `True` if the rule is meet

__Rules:__
1. There are twice as many _energy cells_ in the upper level as in the lower level
2. Each side of the cannon (made of 6 chambers) has 11 _energy cells_

The cannon was originally meant to have $n$ _energy cells_, but Kanoli made it work with $n-3$. What was this number?

BONUS: How do you arrange the _energy cells_?

In [41]:
# RULE 1: There are twice as many energy cells in the upper level as in the lower level
def rule_1(cannon_state):
    layer_1 = cannon_state[0] # lower
    layer_2 = cannon_state[1] # upper
    
    return sum(layer_1) * 2 == sum(layer_2) 

In [42]:
def rule_2(cannon_state):
    layer_1 = cannon_state[0] # lower
    layer_2 = cannon_state[1] # upper

    layer_map = {
        'north': [2,1,0],
        'south': [5,6,7],
        'west' : [0,3,5],
        'east' : [7,4,2],
    }
    
    for side in layer_map.keys():
        indices = layer_map[side]
        
        total = 0
        for index in indices:
            total += layer_1[index] + layer_2[index]
        
        # stop early
        if total != 11:
            return False
    return True

# Solution

In [43]:
ruleset = [rule_1, rule_2]
cannon_gen = create_cannon_state_generator(layers = 2, min_max_range = (1,3))

solutions = []
for cannon_state in cannon_gen:
    passes = True
    for rule in ruleset:
        passes &= rule(cannon_state)
        if not passes: # stop early
            break

    if passes:
        solutions.append(cannon_state)
        print('Found solution!', cannon_state)
print("DONE!")

Found solution! ((1, 1, 1, 1, 1, 1, 1, 2), (3, 2, 3, 2, 1, 3, 1, 3))
Found solution! ((1, 1, 1, 1, 1, 1, 1, 3), (3, 3, 2, 3, 3, 2, 3, 1))
Found solution! ((1, 1, 1, 1, 1, 1, 2, 2), (2, 3, 3, 3, 3, 3, 2, 1))
Found solution! ((1, 1, 1, 1, 1, 1, 2, 2), (3, 2, 3, 3, 3, 2, 3, 1))
Found solution! ((1, 1, 1, 1, 1, 1, 2, 2), (3, 3, 2, 3, 3, 2, 2, 2))
Found solution! ((1, 1, 1, 1, 1, 1, 3, 1), (2, 3, 3, 3, 3, 3, 1, 2))
Found solution! ((1, 1, 1, 1, 1, 1, 3, 1), (3, 2, 3, 3, 3, 2, 2, 2))
Found solution! ((1, 1, 1, 1, 1, 1, 3, 1), (3, 3, 2, 3, 3, 2, 1, 3))
Found solution! ((1, 1, 1, 1, 1, 2, 1, 1), (3, 2, 3, 1, 2, 3, 1, 3))
Found solution! ((1, 1, 1, 1, 1, 2, 1, 2), (2, 3, 3, 3, 3, 2, 3, 1))
Found solution! ((1, 1, 1, 1, 1, 2, 1, 2), (3, 3, 2, 3, 3, 1, 3, 2))
Found solution! ((1, 1, 1, 1, 1, 2, 2, 1), (2, 3, 3, 3, 3, 2, 2, 2))
Found solution! ((1, 1, 1, 1, 1, 2, 2, 1), (3, 2, 3, 3, 3, 1, 3, 2))
Found solution! ((1, 1, 1, 1, 1, 2, 2, 1), (3, 3, 2, 3, 3, 1, 2, 3))
Found solution! ((1, 1, 1, 1, 1, 3

In [44]:
len(solutions)

118

118 solutions!
Certianly duplicates, due to rotated orientations

Let's look into solutions with $n-3$

In [45]:
unique_solutions = {}
for solution in solutions:
    layer_1 = solution[0] # lower
    layer_2 = solution[1] # upper

    total_n = sum(layer_1) + sum(layer_2)
    if total_n not in unique_solutions:
        unique_solutions[total_n] = []
    unique_solutions[total_n].append(solution)

In [47]:
unique_solutions.keys()

dict_keys([27, 30])

We find 2 different kinds of solutions:
1. Solutions that meet the requirments with $n = 27$
1. Solutions that meet the requirments with $n = 30$


In [257]:
unique_solutions[27]

[((1, 1, 1, 1, 1, 1, 1, 2), (3, 2, 3, 2, 1, 3, 1, 3)),
 ((1, 1, 1, 1, 1, 2, 1, 1), (3, 2, 3, 1, 2, 3, 1, 3)),
 ((1, 1, 2, 1, 1, 1, 1, 1), (3, 1, 3, 2, 1, 3, 2, 3)),
 ((2, 1, 1, 1, 1, 1, 1, 1), (3, 1, 3, 1, 2, 3, 2, 3))]

4 solutions have 27 total _energy cells_ (likely duplicates due to rotation)

In [241]:
vis_cannon(unique_solutions[27][0])

                                                                                                                              
          LOWER                UPPER              NORTH            SOUTH              EAST               WEST                 
            n                    n                upper            upper              upper              upper                
      | 1 | 1 | 1 |        | 3 | 2 | 3 |      | 3 | 2 | 3 |    | 3 | 1 | 3 |      | 3 | 1 | 3 |      | 3 | 2 | 3 |
    w | 1 |   | 1 | e    w | 2 |   | 1 | e    | 1 | 1 | 1 |    | 1 | 1 | 2 |      | 2 | 1 | 1 |      | 1 | 1 | 1 |
      | 1 | 1 | 2 |        | 3 | 1 | 3 |          lower            lower              lower              lower              
            s                    s                                                                                            


OBSERVATION:
- The lower level has all 1's except for a single corner 2
- The upper level has 3's on each corner. And 2's in the center, except for the sides around the corner where the lower level has the 2.